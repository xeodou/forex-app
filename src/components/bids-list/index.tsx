import { CurrencyRate } from "lib/forex-api/client";
import { FC } from "react";

export const BidsList: FC<{ rates: CurrencyRate[] }> = ({ rates }) => {
  return (
    <div className="bids-list w-full">
      <header className="flex justify-around uppercase">
        <span>CUR</span>
        <span>Time</span>
        <span>Ask</span>
        <span>Bid</span>
      </header>
      <div className="h-60 overflow-y-auto flex flex-col">
        {rates.map((rate, index) => (
          <ul
            key={rate.time_stamp}
            className={`flex justify-around ${index % 2 === 0 ? "bg-slate-300" : "bg-slate-100 "}`}
          >
            <li>{`${rate.from}/${rate.to}`}</li>
            <li>{new Date(rate.time_stamp).toLocaleString()}</li>
            <li>{rate.ask}</li>
            <li>{rate.bid}</li>
          </ul>
        ))}
      </div>
    </div>
  );
};
