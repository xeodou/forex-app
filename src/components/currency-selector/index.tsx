import { FC } from "react";
import { currencies } from "utils/currencies";

export interface CurrencyHeaderProps {
  // On selected currency changes
  onSelected?: (currency: string) => void;
  // Specific the default selected currency
  value?: string;
}

export const CurrencySelector: FC<CurrencyHeaderProps> = ({ onSelected, value }) => {
  const onChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    onSelected && onSelected(event.target.value);
  };

  return (
    <select
      onChange={onChange}
      value={value}
      className="h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm "
    >
      {currencies.map((currency) => (
        <option key={currency} value={currency}>
          {currency}
        </option>
      ))}
    </select>
  );
};
