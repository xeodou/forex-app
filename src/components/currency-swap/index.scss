.currency-header {
  &--indicator {
    width: 30px;
  }
}
