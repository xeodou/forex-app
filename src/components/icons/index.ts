export * from "./icon-swap";
export * from "./icon-refresh";
export * from './icon-close';
