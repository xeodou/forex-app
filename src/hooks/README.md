## Hooks

The shared hooks for the projects. Please read more if you want to know how does hooks work in React at [here](https://reactjs.org/docs/hooks-intro.html).

### use-currency-pair

This hook that will mantain a [CurrencyPairContext](./use-currency-pair/index.tsx?l=23) and shared the selected currency pair with the other component needs to access the currency pair.

Usage:

```typescript
import { useCurrencyPair } from "hooks";

const [currencyPair, setCurrencyPair] = useCurrencyPair();
```

### use-forex-rates

The hook wrap the forex getRates API into a hook, so you can read the realtime currency rates by React hook.

Usage:

```typescript
import { useForexRates } from "hooks";

const {
  // The realtime forex rates list
  forexRates,
  // The error response
  forexError,
  // A function to reset the data
  reset,
} = useForexRates();
```

### use-notify

This is a global hooks that will matain the notify message context.

Usage:

```typescript
import { useNotify } from "hooks";

const { stack, removeNotifyItem } = useNotify();
```
