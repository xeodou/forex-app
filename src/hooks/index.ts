export * from "./use-currency-pair";
export * from "./use-forex-rates";
