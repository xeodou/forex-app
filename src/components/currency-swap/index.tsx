import { CurrencySelector } from "components/currency-selector";
import { IconSwap } from "components/icons";
import { useCurrencyPair } from "hooks/use-currency-pair";

export const CurrencySwap = ({}) => {
  const [currencyPair, setCurrencyPair] = useCurrencyPair();

  const onFromCurrencyChange = (from: string) => {
    setCurrencyPair([from, currencyPair[1]]);
  };

  const onToCurrencyChange = (to: string) => {
    setCurrencyPair([currencyPair[0], to]);
  };

  const onSwitchClick = () => {
    setCurrencyPair([currencyPair[1], currencyPair[0]]);
  };

  return (
    <div className="flex items-center">
      <CurrencySelector onSelected={onFromCurrencyChange} value={currencyPair[0]} />
      <button
        type="button"
        aria-label="Switch currencies"
        className="font-medium text-sm p-2.5 text-center inline-flex items-center"
        onClick={onSwitchClick}
      >
        <IconSwap></IconSwap>
      </button>
      <CurrencySelector onSelected={onToCurrencyChange} value={currencyPair[1]} />
    </div>
  );
};
