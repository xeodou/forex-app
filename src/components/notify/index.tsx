import { useNotify } from "hooks";
import { Item } from "./_item";

export const Notify = () => {
  const { stack, removeNotifyItem } = useNotify();

  const lastItem = stack[stack.length - 1];

  const onDismiss = (id: number) => {
    removeNotifyItem(id);
  };

  return <div className="notify-container">{lastItem && <Item item={lastItem} onDismiss={onDismiss} />}</div>;
};
