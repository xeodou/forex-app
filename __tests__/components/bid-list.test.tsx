import { act, fireEvent } from "@testing-library/react";
import { render, unmountComponentAtNode } from "react-dom";
import pretty from "pretty";
import { CurrencySwap } from "components/currency-swap";
import { CurrencyPairProvider, defaultCurrencyPair } from "hooks";
import { BidsList } from "components/bids-list";

const mockRates = [
  {
    from: "USD",
    to: "JPY",
    bid: 0.8014505466183891,
    ask: 0.7839126536448034,
    price: 0.7926816001315963,
    time_stamp: "2022-03-12T14:36:44.765Z",
  },
  {
    from: "USD",
    to: "JPY",
    bid: 0.8014505466183891,
    ask: 0.7839126536448034,
    price: 0.7926816001315963,
    time_stamp: "2022-03-12T14:37:44.765Z",
  },
];

describe("bid-list", () => {
  let container: HTMLDivElement;
  let toLocaleTimeString: jest.SpyInstance<string, any>;

  beforeEach(() => {
    toLocaleTimeString = jest.spyOn(Date.prototype, 'toLocaleString');
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null as unknown as HTMLDivElement;
    toLocaleTimeString.mockRestore();
  });

  it("should render the list", () => {
    // Mock the toLocaleString returns to keep the same return across the different CI server.
    toLocaleTimeString.mockReturnValueOnce("3/12/2022, 10:37:44 PM");

    act(() => {
      render(<BidsList rates={mockRates} />, container);
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();

    expect(container.querySelector("header")?.children.length).toBe(4);

    expect(container.querySelectorAll("ul")?.length).toBe(2);
    expect(container.querySelectorAll("ul")[0].classList.contains("bg-slate-300")).toBeTruthy();
    expect(container.querySelectorAll("ul")[1].classList.contains("bg-slate-100")).toBeTruthy();
  });
});
