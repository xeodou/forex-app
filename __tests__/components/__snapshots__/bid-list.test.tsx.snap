// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bid-list should render the list 1`] = `
"<div class=\\"bids-list w-full\\">
  <header class=\\"flex justify-around uppercase\\"><span>CUR</span><span>Time</span><span>Ask</span><span>Bid</span></header>
  <div class=\\"h-60 overflow-y-auto flex flex-col\\">
    <ul class=\\"flex justify-around bg-slate-300\\">
      <li>USD/JPY</li>
      <li>3/12/2022, 2:37:44 PM</li>
      <li>0.7839126536448034</li>
      <li>0.8014505466183891</li>
    </ul>
    <ul class=\\"flex justify-around bg-slate-100 \\">
      <li>USD/JPY</li>
      <li>3/12/2022, 2:37:44 PM</li>
      <li>0.7839126536448034</li>
      <li>0.8014505466183891</li>
    </ul>
  </div>
</div>"
`;
